neo/NeoSystem.cs

```diff
@@ -18,29 +21,109 @@ public class NeoSystem : IDisposable
 
         public ActorSystem ActorSystem { get; } = ActorSystem.Create(nameof(NeoSystem),
             $"akka {{ log-dead-letters = off }}" +
+            $"akka.coordinated-shutdown.phases {{\n" +
+            $"  wait-for-network-shutdown {{\n" +
+            $"    Timeout = 60 s\n" +
+            $"    depends-on = [{CoordinatedShutdown.PhaseBeforeActorSystemTerminate}]\n" +
+            $"  }}\n" +
+            $"  wait-for-neo-shutdown {{\n" +
+            $"    Timeout = 60 s\n" +
+            $"    depends-on = [wait-for-network-shutdown]\n" +
+            $"  }}\n" +
+            $"  actor-system-terminate.depends-on = [wait-for-neo-shutdown]\n" +
+            $"}}" +
             $"blockchain-mailbox {{ mailbox-type: \"{typeof(BlockchainMailbox).AssemblyQualifiedName}\" }}" +
             $"task-manager-mailbox {{ mailbox-type: \"{typeof(TaskManagerMailbox).AssemblyQualifiedName}\" }}" +
             $"remote-node-mailbox {{ mailbox-type: \"{typeof(RemoteNodeMailbox).AssemblyQualifiedName}\" }}" +
             $"protocol-handler-mailbox {{ mailbox-type: \"{typeof(ProtocolHandlerMailbox).AssemblyQualifiedName}\" }}" +
             $"consensus-service-mailbox {{ mailbox-type: \"{typeof(ConsensusServiceMailbox).AssemblyQualifiedName}\" }}");
+        public IActorRef WatchActor { get; }
         public IActorRef Blockchain { get; }
         public IActorRef LocalNode { get; }
         internal IActorRef TaskManager { get; }
         public IActorRef Consensus { get; private set; }
         public RpcServer RpcServer { get; private set; }
 
+        private class ShutdownReason : CoordinatedShutdown.Reason
+        {
+            public static CoordinatedShutdown.Reason Instance = new ShutdownReason();
+
+            private ShutdownReason()
+            {
+            }
+        }
+
+        public class ShutdownWatcher : UntypedActor
+        {
+            public class Watch { public IActorRef ActorRef; };
+            public Dictionary<IActorRef, SemaphoreSlim> ActorSemaphores = new Dictionary<IActorRef, SemaphoreSlim>();
+            private object lockObj = new object();
+            private static ShutdownWatcher singleton;
+            public static ShutdownWatcher Singleton
+            {
+                get
+                {
+                    while (singleton == null) Thread.Sleep(10);
+                    return singleton;
+                }
+            }
+
+            public ShutdownWatcher() { lock (lockObj) singleton = this; }
+
+            protected override void OnReceive(object message)
+            {
+                if (message is Terminated t && ActorSemaphores.TryGetValue(t.ActorRef, out SemaphoreSlim semWait))
+                    semWait.Release(short.MaxValue);
```

Comment:
We could probably call `Context.Unwatch(t.ActorRef)` here also, but perhaps it doesn't matter since we are shutting down anyway."